File: typescript/src/x402-client.ts
import { Keypair } from '@solana/web3.js';
import axios from 'axios';
import * as fs from 'fs';
import { buildExactPaymentProofJsonString } from './pr402-exact-flow';

/**
 * X402Client: high-level buyer for x402 v2 resources that settle via pr402 `exact` rail.
 *
 * Prefer {@link createPay402Fetch} when you already use `fetch` and want a single wrapper.
 */
export class X402Client {
    private payer: Keypair;
    private defaultFacilitator: string;

    /**
     * @param _rpcUrl Reserved for future RPC / balance helpers (optional).
     * @param keypairPath Path to `buyer-keypair.json`
     * @param defaultFacilitatorBaseUrl pr402 origin, e.g. `https://preview.agent.pay402.me`
     */
    constructor(_rpcUrl: string, keypairPath: string, defaultFacilitatorBaseUrl: string) {
        const secretKey = Uint8Array.from(JSON.parse(fs.readFileSync(keypairPath, 'utf-8')));
        this.payer = Keypair.fromSecretKey(secretKey);
        this.defaultFacilitator = defaultFacilitatorBaseUrl.replace(/\/$/, '');
    }

    /** Keypair used for signing payment transactions (same as `createPay402Fetch` `payer`). */
    getPayer(): Keypair {
        return this.payer;
    }

    /**
     * Hit a URL, handle 402 via pr402 `build-exact-payment-tx`, sign, retry with `PAYMENT-SIGNATURE`.
     */
    async buy<T>(url: string, body: unknown, method: 'GET' | 'POST' = 'POST'): Promise<T> {
        console.log(`\x1b[36m[X402] Attempting to access: ${url}\x1b[0m`);

        try {
            const reqMethod = method.toUpperCase() as 'GET' | 'POST';

            const res =
                reqMethod === 'GET'
                    ? await axios.get(url, {
                          ...(body &&
                          typeof body === 'object' &&
                          body !== null &&
                          Object.keys(body as object).length > 0
                              ? { params: body }
                              : {}),
                          validateStatus: (status) => status === 200 || status === 402,
                          headers: { 'Content-Type': 'application/json' },
                      })
                    : await axios.post(url, body, {
                          validateStatus: (status) => status === 200 || status === 402,
                          headers: { 'Content-Type': 'application/json' },
                      });

            if (res.status === 200) {
                return res.data;
            }

            console.log('\x1b[33m[X402] Received 402 Challenge. Settling payment...\x1b[0m');
            const rawHdr = res.headers['payment-required'] as string;
            if (!rawHdr) {
                throw new Error("402 Error: Missing 'Payment-Required' header.");
            }

            const requirements = JSON.parse(Buffer.from(rawHdr, 'base64').toString());

            console.log('\x1b[33m[X402] signing transaction locally...\x1b[0m');
            const finalProof = await buildExactPaymentProofJsonString({
                payer: this.payer,
                requirements,
                defaultFacilitatorBaseUrl: this.defaultFacilitator,
            });

            console.log(
                '\x1b[33m[X402] Final submission with payment proof (Raw JSON)...\x1b[0m',
            );
            const finalRes =
                reqMethod === 'GET'
                    ? await axios.get(url, {
                          ...(body &&
                          typeof body === 'object' &&
                          body !== null &&
                          Object.keys(body as object).length > 0
                              ? { params: body }
                              : {}),
                          headers: {
                              'Content-Type': 'application/json',
                              'PAYMENT-SIGNATURE': finalProof,
                          },
                      })
                    : await axios.post(url, body, {
                          headers: {
                              'Content-Type': 'application/json',
                              'PAYMENT-SIGNATURE': finalProof,
                          },
                      });

            return finalRes.data;
        } catch (error: unknown) {
            const err = error as { message?: string; response?: { data?: unknown } };
            console.error(`\x1b[31m[X402] Error: ${err.message}\x1b[0m`);
            if (err.response) {
                console.error(`\x1b[31mDetail: ${JSON.stringify(err.response.data)}\x1b[0m`);
            }
            throw error;
        }
    }
}

File: python/x402_client.py
import asyncio
import json
import base64
import httpx
from solders.keypair import Keypair
from solders.transaction import VersionedTransaction
from solders.message import to_bytes_versioned

from pr402_defaults import (
    canonical_accepted_for_build,
    facilitator_base_url,
    is_exact_rail_scheme,
)


class X402Client:
    """Buyer-side x402 v2 client for pr402 `exact` rail (build → sign → PAYMENT-SIGNATURE)."""

    def __init__(self, _rpc_url: str, keypair_path: str, default_facilitator_base_url: str):
        self.default_facilitator = default_facilitator_base_url.rstrip("/")
        with open(keypair_path, "r") as f:
            secret = bytes(json.load(f))
            self.payer = Keypair.from_bytes(secret)

    async def buy(self, url: str, body: dict | None = None, method: str = "POST") -> dict:
        print(f"\033[36m[X402] Attempting to access: {url}\033[0m")
        req_method = (method or "POST").upper()

        async with httpx.AsyncClient(timeout=30.0) as client:
            if req_method == "GET":
                if body and len(body.keys()) > 0:
                    res = await client.get(url, params=body)
                else:
                    res = await client.get(url)
            else:
                res = await client.post(url, json=body or {})

            if res.status_code == 200:
                return res.json()

            if res.status_code != 402:
                raise Exception(f"Expected 402, got {res.status_code}: {res.text}")

            print("\033[33m[X402] Received 402 Challenge. Settling payment...\033[0m")
            raw_hdr = res.headers.get("Payment-Required")
            if not raw_hdr:
                raise Exception("Missing 'Payment-Required' header in 402 response.")

            requirements = json.loads(base64.b64decode(raw_hdr))

            accepted = next(
                (
                    a
                    for a in requirements.get("accepts", [])
                    if is_exact_rail_scheme(a.get("scheme"))
                ),
                None,
            )
            if not accepted:
                raise Exception("No supported exact rail in accepts[].")

            extra = accepted.get("extra") or {}
            cap_url = extra.get("capabilitiesUrl") if isinstance(extra, dict) else None
            facilitator_url = facilitator_base_url(
                str(cap_url) if cap_url else None,
                self.default_facilitator,
            )
            build_accepted = canonical_accepted_for_build(accepted)

            build_req = {
                "payer": str(self.payer.pubkey()),
                "accepted": build_accepted,
                "resource": requirements.get("resource"),
            }

            print("\033[33m[X402] Signing transaction locally...\033[0m")
            build_res = await client.post(
                f"{facilitator_url}/api/v1/facilitator/build-exact-payment-tx",
                json=build_req,
            )
            if build_res.status_code != 200:
                raise Exception(f"Facilitator build failed: {build_res.text}")

            build_data = build_res.json()

            tx_bytes = base64.b64decode(build_data["transaction"])
            vtx = VersionedTransaction.from_bytes(tx_bytes)

            required = vtx.message.header.num_required_signatures
            payer_index = None
            for i, key in enumerate(vtx.message.account_keys[:required]):
                if str(key) == str(self.payer.pubkey()):
                    payer_index = i
                    break
            if payer_index is None:
                raise Exception("payer pubkey not found among required signer slots")

            message_bytes = to_bytes_versioned(vtx.message)
            payer_sig = self.payer.sign_message(message_bytes)
            signatures = list(vtx.signatures)
            signatures[payer_index] = payer_sig
            vtx = VersionedTransaction.populate(vtx.message, signatures)

            signed_tx_b64 = base64.b64encode(bytes(vtx)).decode("utf-8")
            verify_body = build_data["verifyBodyTemplate"]
            verify_body["paymentPayload"]["payload"]["transaction"] = signed_tx_b64

            final_proof = json.dumps(verify_body)

            print("\033[33m[X402] Final submission with payment proof (Raw JSON)...\033[0m")
            if req_method == "GET":
                if body and len(body.keys()) > 0:
                    final_res = await client.get(
                        url, params=body, headers={"PAYMENT-SIGNATURE": final_proof}
                    )
                else:
                    final_res = await client.get(url, headers={"PAYMENT-SIGNATURE": final_proof})
            else:
                final_res = await client.post(
                    url, json=body or {}, headers={"PAYMENT-SIGNATURE": final_proof}
                )

            if final_res.status_code != 200:
                raise Exception(f"Final submission failed: {final_res.text}")

            return final_res.json()
